src/main/java/org/prebid/server/bidder/adoppler/AdopplerBidder.java

```diff
@@ -0,0 +1,189 @@
+                        .uri(endpointUrl)
+                        .headers(headers)
+                        .payload(outgoingRequest)
+                        .body(body)
+                        .build());
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+        return Result.of(result, errors);
+    }
+
+    private ExtImpAdoppler parseAndValidateImpExt(Imp imp) {
+        final ExtImpAdoppler extImpAdoppler;
+        try {
+            extImpAdoppler = mapper.mapper().convertValue(imp.getExt(), ADOPPLER_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+        if (StringUtils.isBlank(extImpAdoppler.getAdunit())) {
+            throw new PreBidException(String.format("$.imp.ext.adoppler.adunit required"));
+        }
+        return extImpAdoppler;
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        final int statusCode = httpCall.getResponse().getStatusCode();
+        if (statusCode == HttpResponseStatus.NO_CONTENT.code()) {
+            return Result.of(Collections.emptyList(), Collections.emptyList());
+        } else if (statusCode == HttpResponseStatus.BAD_REQUEST.code()) {
+            return Result.emptyWithError(BidderError.badInput("bad request"));
+        } else if (statusCode != HttpResponseStatus.OK.code()) {
+            return Result.emptyWithError(BidderError.badServerResponse(String.format("Unexpected HTTP status %s.",
+                    statusCode)));
+        }
+
+        final BidResponse bidResponse;
+        try {
+            bidResponse = decodeBodyToBidResponse(httpCall);
+        } catch (PreBidException e) {
+            return Result.emptyWithError(BidderError.badServerResponse(String.format("invalid body: %s",
+                    e.getMessage())));
+        }
+
+        final Map<String, BidType> impTypes = new HashMap<>();
+        for (Imp imp : bidRequest.getImp()) {
+            if (impTypes.get(imp.getId()) != null) {
+                return Result.emptyWithError(BidderError.badInput(String.format("duplicate $.imp.id %s", imp.getId())));
+            }
+            if (imp.getBanner() != null) {
+                impTypes.put(imp.getId(), BidType.banner);
+            } else if (imp.getVideo() != null) {
+                impTypes.put(imp.getId(), BidType.video);
+            } else if (imp.getAudio() != null) {
+                impTypes.put(imp.getId(), BidType.audio);
+            } else if (imp.getXNative() != null) {
+                impTypes.put(imp.getId(), BidType.xNative);
+            } else {
+                return Result.emptyWithError(BidderError.badInput(String.format(
+                        "one of $.imp.banner, $.imp.video, $.imp.audio and $.imp.native field required")));
+            }
```

Comment:
I'd probably extracted this into separate method

Reply:
I think It is not good idea due to return statements